File: scripts/generate-images.mjs
#!/usr/bin/env node
/**
 * Generate branded site imagery with OpenAI gpt-image-1.
 *
 * Usage:
 *   OPENAI_API_KEY=sk-... node scripts/generate-images.mjs           # generate missing
 *   OPENAI_API_KEY=sk-... node scripts/generate-images.mjs --force   # regenerate all
 *   OPENAI_API_KEY=sk-... node scripts/generate-images.mjs hero-home  # one image by id
 *
 * Output: public/generated/<id>.png  (referenced via lib/images.ts)
 */
import { readFileSync, existsSync, mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "..");
const OUT_DIR = join(ROOT, "public", "generated");

// Supports either standard OpenAI or Azure OpenAI (services.ai.azure.com/openai/v1).
const KEY = process.env.OPENAI_API_KEY || process.env.AZURE_OPENAI_KEY;
const AZURE_ENDPOINT = (process.env.AZURE_OPENAI_ENDPOINT || "").replace(/\/$/, "");
const AZURE_DEPLOYMENT = process.env.AZURE_OPENAI_IMAGE_DEPLOYMENT || "gpt-image-2";
const IS_AZURE = !!AZURE_ENDPOINT;
if (!KEY) {
  console.error("✗ Set OPENAI_API_KEY (or AZURE_OPENAI_KEY + AZURE_OPENAI_ENDPOINT) first.");
  process.exit(1);
}
const API_URL = IS_AZURE ? `${AZURE_ENDPOINT}/images/generations` : "https://api.openai.com/v1/images/generations";
const MODEL = IS_AZURE ? AZURE_DEPLOYMENT : "gpt-image-1";

const force = process.argv.includes("--force");
const onlyId = process.argv.find((a) => !a.startsWith("-") && !a.endsWith(".mjs") && !a.includes("node"));

const manifest = JSON.parse(readFileSync(join(__dirname, "image-manifest.json"), "utf8"));
const baseStyle = manifest.style;
let images = manifest.images;
if (onlyId) images = images.filter((i) => i.id === onlyId);

if (!existsSync(OUT_DIR)) mkdirSync(OUT_DIR, { recursive: true });

async function generate(img) {
  const dest = join(OUT_DIR, `${img.id}.png`);
  if (existsSync(dest) && !force) {
    console.log(`• skip ${img.id} (exists)`);
    return;
  }
  const prompt = `${img.prompt}\n\nStyle: ${baseStyle}`;
  process.stdout.write(`… generating ${img.id} (${img.size}) `);
  const headers = IS_AZURE
    ? { "api-key": KEY, "Content-Type": "application/json" }
    : { Authorization: `Bearer ${KEY}`, "Content-Type": "application/json" };
  const res = await fetch(API_URL, {
    method: "POST",
    headers,
    body: JSON.stringify({ model: MODEL, prompt, size: img.size, n: 1 })
  });
  if (!res.ok) {
    console.log("✗");
    console.error(`  ${res.status}: ${await res.text()}`);
    return;
  }
  const json = await res.json();
  const b64 = json.data?.[0]?.b64_json;
  if (!b64) {
    console.log("✗ no image data");
    return;
  }
  writeFileSync(dest, Buffer.from(b64, "base64"));
  console.log("✓");
}

console.log(`Generating ${images.length} image(s) → public/generated/`);
for (const img of images) {
  try {
    await generate(img);
  } catch (e) {
    console.error(`✗ ${img.id}: ${e.message}`);
  }
}
console.log("Done. Commit the PNGs in public/generated/ to ship them.");

File: scripts/generate-images.py
#!/usr/bin/env python3
"""Generate Hanuone brand imagery via Azure gpt-image-2 -> public/generated/.

Run:  IMG_KEY="<azure-key>" python3 scripts/generate-images.py [id1 id2 ...]
Auth that works for this resource: BOTH api-key + Bearer headers, ?api-version=preview.
"""
import os, sys, json, time, base64, urllib.request, urllib.error
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "public" / "generated"
OUT.mkdir(parents=True, exist_ok=True)

KEY = os.environ.get("IMG_KEY") or os.environ.get("OPENAI_API_KEY")
BASE = os.environ.get("IMG_BASE", "https://free-credits-openai-pro-resource.services.ai.azure.com/openai/v1")
MODEL = os.environ.get("IMG_MODEL", "gpt-image-2")
if not KEY:
    sys.exit("Set IMG_KEY")

manifest = json.loads((ROOT / "scripts" / "image-manifest.json").read_text())
style = manifest["style"]
jobs = manifest["images"]
only = sys.argv[1:]
if only:
    jobs = [j for j in jobs if j["id"] in only]

URL = f"{BASE}/images/generations?api-version=preview"

def gen(job):
    prompt = f"{job['prompt']}\n\nStyle: {style}"
    body = json.dumps({"model": MODEL, "prompt": prompt, "n": 1, "size": job["size"]}).encode()
    req = urllib.request.Request(URL, data=body, method="POST")
    req.add_header("api-key", KEY)
    req.add_header("Authorization", f"Bearer {KEY}")
    req.add_header("Content-Type", "application/json")
    with urllib.request.urlopen(req, timeout=180) as resp:
        data = json.loads(resp.read())
    b64 = data["data"][0]["b64_json"]
    (OUT / f"{job['id']}.png").write_bytes(base64.b64decode(b64))

ok = 0
for j in jobs:
    for attempt in range(1, 6):
        try:
            print(f"… {j['id']} ({j['size']}) ", end="", flush=True)
            gen(j)
            print("✓")
            ok += 1
            time.sleep(12)  # pace under per-minute limit
            break
        except urllib.error.HTTPError as e:
            code = e.code
            if code == 429:
                print("429, waiting 25s")
                time.sleep(25); continue
            print(f"✗ HTTP {code}: {e.read()[:160].decode('utf-8','replace')}")
            time.sleep(4)
            if attempt == 5: break
        except Exception as e:
            print(f"✗ {str(e)[:140]}")
            time.sleep(4)
            if attempt == 5: break

print(f"\nGenerated {ok}/{len(jobs)} -> public/generated/")
